### PR TITLE
Allow FMP trace event to appear slightly before the FCP

### LINKED ISSUE #1253: Cannot read property 'ts' of undefined
After the chrome extension update, I keep getting the error whenever I try to generate the report.
I don't see this error in my application when run in the browser.

screen shot 2016-12-22 at 8 46 08 pm

### PR DESCRIPTION
**Pull request by paulirish**
This is the core problem of #1253.

As @wardpeet deduced, we are hitting these because sometimes FMP trace events are added before FCP

This now allows there to be 32ms of tolerance between the two, as we found traces where FMP is just 0.7ms before FCP.

Fixes #1253.

### CODE REVIEW

**Comment by paulirish** [comment]
@googlebot cool thanks

**Comment by brendankenny** [comment]
you have some camel case eslint errors

File: lighthouse-core/audits/first-meaningful-paint.js [hunk 1]
```diff
@@ -133,9 +136,15 @@ class FirstMeaningfulPaint extends Audit {
...
     const navigationStart = frameEvents.filter(e =>
         e.name === 'navigationStart' && e.ts < firstFCP.ts).pop();
-    // FMP will follow at/after the FCP
+    // fMP will follow at/after the FCP, though we allow some timestamp tolerance
     const firstMeaningfulPaint = frameEvents.find(e =>
-        e.name === 'firstMeaningfulPaint' && e.ts >= firstFCP.ts);
+        e.name === 'firstMeaningfulPaint' && e.ts >= (firstFCP.ts - fCPfMP_TOLERANCE));
+
+    // Sometimes fMP is triggered before fCP
+    // (this happends when above the fold is being swapped with JS)
```
Comment (brendankenny) [inline lighthouse-core/audits/first-meaningful-paint.js]: `happens`

(also not sure what this means. Fold?)

File: lighthouse-core/audits/first-meaningful-paint.js [hunk 2]
```diff
@@ -133,9 +136,15 @@ class FirstMeaningfulPaint extends Audit {
     // Our navStart will be the latest one before fCP.
     const navigationStart = frameEvents.filter(e =>
         e.name === 'navigationStart' && e.ts < firstFCP.ts).pop();
-    // FMP will follow at/after the FCP
+    // fMP will follow at/after the FCP, though we allow some timestamp tolerance
     const firstMeaningfulPaint = frameEvents.find(e =>
-        e.name === 'firstMeaningfulPaint' && e.ts >= firstFCP.ts);
+        e.name === 'firstMeaningfulPaint' && e.ts >= (firstFCP.ts - fCPfMP_TOLERANCE));
```
Comment (brendankenny) [inline lighthouse-core/audits/first-meaningful-paint.js]: if there's an FMP event a few ms before FCP (within tolerance) *and* one after FCP, is the first one always better here?

File: lighthouse-core/test/fixtures/traces/preactjs.com_ts_of_undefined.json [hunk 3]
```diff
@@ -0,0 +1,77 @@
+{
```
Comment (brendankenny) [inline lighthouse-core/test/fixtures/traces/preactjs.com_ts_of_undefined.json]: hooray for minimal trace fixtures :)

File: lighthouse-core/audits/first-meaningful-paint.js (hunk 2 above)
Comment (paulirish) [inline lighthouse-core/audits/first-meaningful-paint.js]: this shouldn't happen as we filter by frame. but if so, yeah sure. :)

File: lighthouse-core/audits/first-meaningful-paint.js [hunk 4]
```diff
@@ -133,9 +139,20 @@ class FirstMeaningfulPaint extends Audit {
...
-        e.name === 'firstMeaningfulPaint' && e.ts >= firstFCP.ts);
+        e.name === 'firstMeaningfulPaint' && e.ts >= (firstFCP.ts - FCPFMP_TOLERANCE));
+
+    // Sometimes fMP is triggered before fCP
+    if (!firstMeaningfulPaint) {
+      throw new Error('No usable `firstMeaningfulPaint` event found in trace');
+    }
+
+    // Sometimes fMP is triggered before fCP
+    // (this happends when above the fold is being swapped with JS)
```
Comment (brendankenny) [inline lighthouse-core/audits/first-meaningful-paint.js]: rebase or something got mixed up here

**Comment by tdresser** [comment]
Woah, FCP should never happen after FMP.
If that's happening, there's a bug.

**Comment by brendankenny** [comment]
@tdresser there's an example trace with FMP before FCP in [www.nfl.com_2017-01-04_09-44-25-0.zip](https://github.com/GoogleChrome/lighthouse/files/685458/www.nfl.com_2017-01-04_09-44-25-0.zip) from https://github.com/GoogleChrome/lighthouse/issues/1253#issuecomment-270438352

**Comment by wardpeet** [comment]
@tdresser here is another one
[whoistheorchid.com](https://drive.google.com/open?id=0B9_w9RCPQgsMTUxzV2JXN1l0TjA)

**Pull request merged**